bench: close wallets after migration

l0rinc opened the pull request:
The low-priority `WalletMigration` benchmark existed for some time but was never run automatically in our CI.

Although the failure first surfaced on Windows as a hang during temporary directory cleanup, it could also be reproduced on Linux and macOS when forcing multiple iterations (e.g. via a long `--min-time`).

### Root causes

1. **Leaked open wallets on Windows**
   `MigrateLegacyToDescriptor` produces two new descriptor wallets (the primary spendable wallet and a companion watch‑only wallet). Without unloading them, their database files remained open in the `WalletContext`, blocking directory removal and hanging the test harness.

<details><summary>Details</summary>

```bash
what():  filesystem error: cannot remove all: The process cannot access the file because it is being used by another process [C:\Users\RUNNER\~1\AppData\Local\Temp\test_common bitcoin\WalletMigration\d8ffd89a7700ce01c31f] [C:\Users\RUNNER~1\AppData\Local\Temp\test_common bitcoin\WalletMigration\d8ffd89a7700ce01c31f\regtest\wallet.dat]
```

</details>

2. **Undefined behavior on repeated runs**
   The benchmark body calls `std::move(wallet)`, invalidating the local `wallet` pointer. Running more than one iteration causes a use-after-move by the sanitizers.

<details><summary>Details</summary>

```bash
error: bench_bitcoin 0x00067927: DW_TAG_member '_M_local_buf' refers to type 0x00000000000b3ba7 which extends beyond the bounds of 0x0006791d
* thread #1, name = 'b-test', stop reason = signal SIGSEGV: address not mapped to object (fault address: 0xc8)
  * frame #0: 0x00005555556a3f33 bench_bitcoin`... basic_string<char>::length(this=<unavailable>) const at basic_string.h:1079:16
```

</details>

### Fixes

- **Automatic wallet teardown**
  Wrap the benchmark in a `MakeWalletLoader` (owning a `WalletContext`), so that both migrated wallets are unloaded when the loader goes out of scope, eliminating any lingering open files.

- **Re-enable benchmarks in CI**
  Drop the temporary filter in GitHub Actions. The `-sanity-check` run already executes each benchmark once, so `WalletMigration` now runs automatically without hangs or crashes.

- **Single iteration**
  Configure the microbenchmark with `.epochs(1).epochIterations(1)`, ensuring the migration code runs exactly once and avoiding use-after-move.

No measurable change in benchmark performance.

hebasto commented on src/bench/wallet_migration.cpp:
```diff
@@ -68,6 +68,9 @@ static void WalletMigration(benchmark::Bench& bench)
         assert(res->wallet);
         assert(res->watchonly_wallet);
+
+        res->wallet->Close();
+        res->watchonly_wallet->Close();
```
Should the fact that the migrated wallet databases remain open be documented in `MigrateLegacyToDescriptor`?

l0rinc replied:
I needed to close both for the build to pass (`res->wallet->Close()` alone wasn't enough).
Maybe it passes with only `res->watchonly_wallet->Close()` - I can check if you think it makes sense.

hebasto replied:
I wonder why `res->watchonly_wallet.use_count() == 3`? Is this the expected behavior of the `MigrateLegacyToDescriptor()` function?

l0rinc replied:
Maybe @furszy knows.

hebasto replied:
Unfortunately, this patch does not fix the test case posted [here](https://github.com/bitcoin/bitcoin/issues/32291#issuecomment-2816658407).

l0rinc replied:
> Maybe it passes with only res->watchonly_wallet->Close() - I can check, if you think it makes sense

No, we do need to close both: https://github.com/l0rinc/bitcoin/actions/runs/14548535133/job/40817048030?pr=9

l0rinc replied:
> Unfortunately, this patch does not fix the test case posted https://github.com/bitcoin/bitcoin/issues/32291#issuecomment-2816658407.

I can reproduce the failure with
```bash
cmake --preset=libfuzzer -DBUILD_FOR_FUZZING=OFF -DBUILD_BENCH=ON && cmake --build build_fuzz -j$(nproc) && build_fuzz/bin/bench_bitcoin -filter='WalletMigration' --min-time=10000
```

I pushed a new version based on https://github.com/bitcoin/bitcoin/blob/master/src/bench/wallet_create.cpp#L54 which calls RemoveWallet instead - @hebasto, can you please check if this fixes it for you as well?

l0rinc replied:
Pushed again, please recheck

l0rinc committed:
bench: clean up migrated descriptor wallets via loader teardown

`MigrateLegacyToDescriptor` returns both a spendable descriptor wallet and a watch‑only wallet.
If these remain attached, their files stay open and on Windows this can hang CI when removing the test directory.

By constructing them via `MakeWalletLoader` (which owns the `WalletContext`), both wallets are automatically unloaded when the loader is destroyed at the end.
This ensures no lingering handles or resource leaks when running the benchmark on CI with `-sanity-check`.

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

hebasto commented:
In commit 86646448770888f66c446eacdd47d1a62d5552eb, "... we're only running 1 epochs anyway" -- where does this assumption comes from?

l0rinc commented:
> where does this assumption comes 

> bench.epochs(/*numEpochs=*/1)

hebasto commented:
Tested 86646448770888f66c446eacdd47d1a62d5552eb on Ubuntu 24.04:
```
$ ./build/bin/bench_bitcoin -filter=WalletMigration -min-time=100000
Warning, results might be unstable:
* CPU frequency scaling enabled: CPU 0 between 800.0 and 4,800.0 MHz
* CPU governor is 'powersave' but should be 'performance'
* Turbo is enabled, CPU frequency will fluctuate

Recommendations
* Use 'pyperf system tune' before benchmarking. See https://github.com/psf/pyperf
Segmentation fault (core dumped)
```

l0rinc commented:
> Segmentation fault (core dumped)

Can you show me the full build command please, `cmake --preset=libfuzzer -DBUILD_FOR_FUZZING=OFF -DBUILD_BENCH=ON && cmake --build build_fuzz -j$(nproc) && build_fuzz/bin/bench_bitcoin -filter='WalletMigration' --min-time=10000` passes for me on Ubuntu.

l0rinc committed:
ci: re-enable all benchmark runs

Drop the temporary `-filter` that excluded the `WalletMigration` benchmark.

hebasto commented:
> > Segmentation fault (core dumped)
> 
> Can you show me the full build command please, `cmake --preset=libfuzzer -DBUILD_FOR_FUZZING=OFF -DBUILD_BENCH=ON && cmake --build build_fuzz -j$(nproc) && build_fuzz/bin/bench_bitcoin -filter='WalletMigration' --min-time=10000` passes for me on Ubuntu.

```
$ rm -rf build && cmake -B build -DBUILD_BENCH=ON
$ cmake --build build -t bench_bitcoin
$ ./build/bin/bench_bitcoin -filter=WalletMigration -min-time=100000
Warning, results might be unstable:
* CPU frequency scaling enabled: CPU 0 between 800.0 and 4,800.0 MHz
* CPU governor is 'powersave' but should be 'performance'
* Turbo is enabled, CPU frequency will fluctuate

Recommendations
* Use 'pyperf system tune' before benchmarking. See https://github.com/psf/pyperf
Segmentation fault (core dumped)
```

l0rinc committed:
bench: ensure wallet migration benchmark runs exactly once

The migration benchmark crashes if run more than once, because of `std::move(wallet)` and leaves subsequent iterations in an undefined state - avoiding `UndefinedBehaviorSanitizer` null‑dereference error.

mzumsande reviewed (commented):
Would be good to explain 1.) Why it failed before and 2.) How this is fixed 
Looking at the commits, it seems that there are two contradictory things going on:
Making sure that the benchmarks could have multiple invocations (commit 1 and 3), and making sure that it actually only runs once (commit 2).

l0rinc commented:
Thanks, I have updated the description, let me know if it answered your questions.

Short answer anyway: we have to clean up regardless of the execution count - and the benchmark seems to have been designed for a single execution anyway, otherwise recreating the wallet might have been too costly.

Let me know if you have a better idea.

mzumsande commented:
Thanks, I didn't realize that there were multiple independent issues - makes sense now, Concept ACK.

furszy commented:
> On Windows, benchmark runs sporadically hung because the descriptor wallet and its companion watch-only wallet were never unloaded, remaining registered in the global WalletContext. This prevented the test harness from cleaning up its temporary directory.

Hmm, haven't dug deeper but this does not seem to be accurate. The `WalletContext` is not global. It is created at the beginning of the benchmark, just after creating the test context (which is the one that calls to the `fs::removal_all` that crashes during destruction).

That said, I think you could just construct a wallet loader instead of manually removing the wallets in the code we want to benchmark. The loader internally creates a `WalletContext` and unloads all wallets during destruction. E.g.
```diff
diff --git a/src/bench/wallet_migration.cpp b/src/bench/wallet_migration.cpp
--- a/src/bench/wallet_migration.cpp
+++ b/src/bench/wallet_migration.cpp
@@ -5,6 +5,7 @@
 #include <bench/bench.h>
 #include <kernel/chain.h>
 #include <interfaces/chain.h>
+#include <interfaces/wallet.h>
 #include <node/context.h>
 #include <test/util/mining.h>
 #include <test/util/setup_common.h>
@@ -21,9 +22,7 @@
 {
     const auto test_setup = MakeNoLogFileContext<TestingSetup>();

-    WalletContext context;
-    context.args = &test_setup->m_args;
-    context.chain = test_setup->m_node.chain.get();
+    std::unique_ptr<interfaces::WalletLoader> loader = MakeWalletLoader(*test_setup->m_node.chain, test_setup->m_args);
 
     // Number of imported watch only addresses
     int NUM_WATCH_ONLY_ADDR = 20;
@@ -63,6 +62,7 @@
         batch.WriteKey(pubkey, key.GetPrivKey(), CKeyMetadata());
     }
 
+    WalletContext& context{*loader->context()};
     bench.epochs(/*numEpochs=*/1).run([&context, &wallet] {
         util::Result<MigrationResult> res = MigrateLegacyToDescriptor(std::move(wallet), /*passphrase=*/"", context, /*was_loaded=*/false);
         assert(res);
```
Also, if you want to double-ensure unload before test context destruction, you could just manually delete the loader at the end of the benchmark function (outside the benchmark lambda).

> The benchmark harness executed the benchmark lambda multiple times unless explicitly told not to. After the first invocation moved the wallet (std::move(wallet)), subsequent iterations operated on a moved-from object, leading to undefined behavior. While re-creating the wallet in each iteration would have worked, it defeats the purpose of measuring migration performance. Since the benchmark is already slow and likely intended to run only once, the iteration count was capped at one.

When we run the benchmark in the CI, we provide the `-sanity-check` flag, which already sets `bench.epochs(1).epochIterations(1)`. While it’s good to also set this inside the benchmark for standalone runs, this change shouldn't be related to the CI issue.

shahsb reviewed (commented):
Concept ACK

achow101 commented:
Is a benchmark that runs only once actually useful?

l0rinc commented:
Not sure, we can decide to delete it as well - I'll push tomorrow a fixed version based on @furszy's recommendation (but have to retest it first after my benchmarking servers free up again).
@furszy, is this still relevant or would deleting it be simpler?

achow101 commented:
I guess it is useful as a spot check, but the accuracy is not very good.

Concept ACK

Tested on Windows and it worked. Was also able to reproduce failure 3, but don't know how to reproduce failure 2.

furszy commented:
> Not sure, we can decide to delete it as well - I'll push tomorrow a fixed version based on @furszy's recommendation (but have to retest it first after my benchmarking servers free up again).
> @furszy, is this still relevant or would deleting it be simpler?

I don't think anyone is planning to touch the legacy wallet code anymore, and the migration procedure no longer takes an hour either (which was the reason behind its creation). So I'd say it's no longer relevant as a benchmark.

Also, agree with achow. Since the introduction of the read-only BDB class, this benchmark no longer covers db reads either. So it's not realistically accurate. I believe we've kept it merely as a sanity check because it's the only test that will survive the annihilation of the legacy wallet code.

l0rinc commented:
Thanks for the suggestions @furszy, @achow101 and @maflcko.
Updated the commits to use `MakeWalletLoader` instead of manually closing, and moved the `epochIterations(1)` after the CI change to obviate that the CI already does that.

TheCharlatan commented:
> I believe we've kept it merely as a sanity check because it's the only test that will survive the annihilation of the legacy wallet code.

Sounds like this should be an actual test then instead of a benchmark, no?

furszy reviewed (commented):
utACK cad39f86fb5a81

> Sounds like this should be an actual test then instead of a benchmark, no?

I'd keep it as is for now and drive all our focus towards #31250. It would be nice to cut down the tree first - we can pick up any leftover later.

l0rinc commented on src/bench/wallet_migration.cpp:
```diff
@@ -63,12 +61,13 @@ static void WalletMigration(benchmark::Bench& bench)
-        util::Result<MigrationResult> res = MigrateLegacyToDescriptor(std::move(wallet), /*passphrase=*/"", context, /*was_loaded=*/false);
-        assert(res);
-        assert(res->wallet);
-        assert(res->watchonly_wallet);
```
you mean like this, i.e. one less space for the block?
```C++
bench.epochs(/*numEpochs=*/1).epochIterations(/*numIters=*/1) // run the migration exactly once
    .run([&] {
```
What do you mean exactly by "would reduce the whitespace churn of this commit", we still need to indent it (and the methods are aligned this way, I like that more). You can convince me otherwise, but if it's a `nit`, I'd prefer the current one.

Or if you meant
```C++
bench.epochs(/*numEpochs=*/1).epochIterations(/*numIters=*/1).run([&] { // run the migration exactly once
```
, in which case there wouldn't be any new indentation, but the current comment prevent `clang-format` from doing that automatically.

Let me know if you have strong preference here.

l0rinc replied:
I'm all for consistency - but is this not what you're getting when formatting the latest commit (except for that 1 space difference I mentioned above)?
If I need to touch it again, I'll put the run to the previous line to minimize the diff (though I consider this to be slightly more readable).

hebasto reviewed (approved):
ACK cad39f86fb5a81f0e3b5116e8e989bab8af89718, tested on Ubuntu 25.04.

furszy commented:
> I believe we've kept it merely as a sanity check because it's the only test that will survive the annihilation of the legacy wallet code.

Best bench-test ever. Just caught a real bug in #31423.
